### PR TITLE
Fix mps unary op issue on non densely stored tensors

### LINKED ISSUE #105284: unary ops on strided int tensors fails on MPS
### 🐛 Describe the bug

I encountered this problem when I pulled main and saw a new failing test around `torch.vander` in my pr #104688 (the pr technically enables `torch.vander` as a side effect which is why I was testing it). However everything below is about main to isolate it from my changes.

```python
import torch
x = torch.tensor([1, 2, 3, 5, 6], device="mps", dtype=torch.int16)
x.unsqueeze(-1).expand(x.shape + (3,)).cumsum(-1)
```

currently results in:
```
Python 3.11.3 (main, May 15 2023, 18:01:31) [Clang 14.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
x = torch.tensor([1, 2, 3, 5, 6], device="mps", dtype=torch.int16)
x.unsqueeze(-1).expand(x.shape + (3,)).cumsum(-1)
>>> x = torch.tensor([1, 2, 3, 5, 6], device="mps", dtype=torch.int16)
>>> x.unsqueeze(-1).expand(x.shape + (3,)).cumsum(-1)
/AppleInternal/Library/BuildRoots/97f6331a-ba75-11ed-a4bc-863efbbaf80d/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphExecutable.mm:1377: failed assertion `Incompatible element type for parameter at index 0, mlir module expected element type si16 but received si64'
zsh: abort      python
```

The same thing happens with other unary operators as well:
```python
import torch
x = torch.tensor([1, 2, 3, 5, 6], device="mps", dtype=torch.int16)
x.unsqueeze(-1).expand(x.shape + (3,)).logit(-1)
```
results in
```
Python 3.11.3 (main, May 15 2023, 18:01:31) [Clang 14.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> x = torch.tensor([1, 2, 3, 5, 6], device="mps", dtype=torch.int16)
>>> x.unsqueeze(-1).expand(x.shape + (3,)).logit(-1)
/AppleInternal/Library/BuildRoots/97f6331a-ba75-11ed-a4bc-863efbbaf80d/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphExecutable.mm:1377: failed assertion `Incompatible element type for parameter at index 0, mlir module expected element type si16 but received f32'
zsh: abort      python
```

I believe this error is due to [this line](https://github.com/pytorch/pytorch/pull/100765/files#:~:text=self_%20%3D%20at%3A%3Aempty_like(output)%3B) in this [recent pr](https://github.com/pytorch/pytorch/pull/100765/files). By changing the line:
```
self_ = at::empty_like(output);
```
to 
```
self_ = at::empty_like(self);
```
I resolved this issue. The reason this is a problem is because the logic in cumsum currently assumes the input scalar type is represented by the scalar type of the input tensor not the output tensor. See [here](https://github.com/pytorch/pytorch/blob/efdabbff06611078de275d085f75f724a8059740/aten/src/ATen/native/mps/operations/UnaryOps.mm#L401C17-L401C17). This logic makes me think that the previously mentioned line:
```
self_ = at::empty_like(output);
```
is incorrect. However I don't know the intention behind adding that line and if it is correct I can just alter the cumsum logic in my pr #104688. So I wanted to file a separate issue to discuss this specifically without tying it to #104688

### Versions

Collecting environment information...
PyTorch version: 2.1.0.dev20230526
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.3.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.24.2
Libc version: N/A

Python version: 3.9.12 (main, Apr  5 2022, 01:52:34)  [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-13.3.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.1.0.dev20230526
[pip3] torchaudio==2.1.0.dev20230527
[pip3] torchvision==0.16.0.dev20230526
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] torch                     2.1.0.dev20230526          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230527          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230526          pypi_0    pypi

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by Mr4k**
This pr fixes a bug where non densely stored tensors were not converted to the dense tensors of the correct scalar type in the mps `unary_op` helper function

Fixes https://github.com/pytorch/pytorch/issues/105284

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/UnaryOps.mm [hunk 1]
```diff
@@ -81,10 +81,10 @@ void unary_op(const Tensor& self,
       newCachedGraph->outputTensor_ = unaryBlock(mpsGraph, castTensor);
     });
 
-    // If self is densely mapped in storage, create a dense output-like representation
+    // If self is non-densely mapped in storage, create a dense representation
     at::Tensor self_;
     if (!is_dense_in_storage(self)) {
-      self_ = at::empty_like(output);
+      self_ = at::empty_like(self);
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/UnaryOps.mm:87]: This is wrong (and I hope tests will capture this): it needs to re-stride self to be output like, but preserve dtype (and this is where the problem is)

Comment (malfet) [inline aten/src/ATen/native/mps/operations/UnaryOps.mm:87]:
```suggestion
      self_ = at::empty_like(output, self.scalar_type());
```

File: aten/src/ATen/native/mps/operations/UnaryOps.mm [hunk 2]
```diff
@@ -81,10 +81,10 @@ void unary_op(const Tensor& self,
       newCachedGraph->outputTensor_ = unaryBlock(mpsGraph, castTensor);
     });
 
-    // If self is densely mapped in storage, create a dense output-like representation
+    // If self is non-densely mapped in storage, create a dense representation
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/UnaryOps.mm:84]:
```suggestion
    // If self is non-densely mapped in storage, create a dense output-like representation
```

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
@Mr4k fixed issue referenced (and applied suggestion, it's important to keep self cast to be output-like), when MPS builds/tests are green will merge it

File: aten/src/ATen/native/mps/operations/UnaryOps.mm (hunk 1 above)
Comment (Mr4k) [reply aten/src/ATen/native/mps/operations/UnaryOps.mm]: great catch! Yea I knew the issue was the dtype but I didn't think about the striding being copied over, I learned something about tensors today

**Comment by Mr4k** [comment]
Thanks for the help and the quick response!

**Pull request closed without merging**